python/ray/data/impl/push_based_shuffle.py

```diff
@@ -0,0 +1,388 @@
+
+        map_fn = self._map_partition
+        merge_fn = self._merge
+
+        def map_partition(*args, **kwargs):
+            return map_fn(self.map, *args, **kwargs)
+
+        def merge(*args, **kwargs):
+            return merge_fn(self.reduce, *args, **kwargs)
+
+        shuffle_map = cached_remote_fn(map_partition)
+        shuffle_merge = cached_remote_fn(merge)
+        shuffle_reduce = cached_remote_fn(self.reduce)
+
+        # Compute all constants used for task scheduling.
+        cpu_map = PushBasedShufflePlan._get_cluster_cpu_map()
+        num_cpus_total = sum(v for v in cpu_map.values())
+        task_parallelism = min(num_cpus_total, input_num_blocks)
+        # Total number of merge tasks.
+        num_merge_tasks = math.ceil(input_num_blocks / merge_factor)
+        # Number of map-merge rounds.
+        num_rounds = math.ceil((input_num_blocks + num_merge_tasks) / task_parallelism)
+        # M: Number of map tasks in one map-merge round.
+        num_map_tasks_per_round = math.ceil(input_num_blocks / num_rounds)
+        # N: Number of merge tasks in one map-merge round. Each map_partition
+        # task will send one output to each of these merge tasks.
+        num_merge_tasks_per_round = math.ceil(num_merge_tasks / num_rounds)
+        # Scheduling args for assigning merge tasks to nodes. We use node-affinity
+        # scheduling here to colocate merge tasks that output to the same
+        # reducer.
+        merge_task_args = self._compute_merge_task_options(
+            num_merge_tasks_per_round, merge_factor, cpu_map
+        )
+
+        # Intermediate results for the map-merge stage.
+        map_results = []
+        # ObjectRef results from the last round of tasks. Used to add
+        # backpressure during pipelining of map and merge tasks.
+        last_map_metadata_results = []
+        last_merge_metadata_results = []
+        mapper_idx = 0
+        # Preemptively clear the blocks list since we will incrementally delete
+        # the last remaining references as we submit the dependent map tasks
+        # during the map-merge stage.
+        if clear_input_blocks:
+            input_blocks.clear()
+
+        # Final outputs from the map-merge stage.
+        # This is a map from merge task index to a nested list of merge results
+        # (ObjectRefs). Each merge task index corresponds to a partition of P
+        # final reduce tasks.
+        all_merge_results = defaultdict(list)
+        shuffle_map_metadata = []
+        shuffle_merge_metadata = []
+        map_bar = ProgressBar("Shuffle Map", position=0, total=input_num_blocks)
+        # Execute the map-merge stage. This submits tasks in rounds of M map
+        # tasks and N merge tasks each. Task execution between map and merge is
+        # pipelined, so that while executing merge for one round of inputs, we
+        # also execute the map tasks for the following round.
+        while input_blocks_list:
```

Comment:
can we wrap this block into a helper function as well, for readability? e.g., `execute_map_stage`